libpius/path.py

```diff
@@ -0,0 +1,150 @@
+
+    if path is None:
+        path = os.environ.get("PATH", os.defpath)
+    if not path:
+        return None
+    path = path.split(os.pathsep)
+
+    if sys.platform == "win32":
+        # The current directory takes precedence on Windows.
+        if not os.curdir in path:
+            path.insert(0, os.curdir)
+
+        # PATHEXT is necessary to check on Windows.
+        pathext = os.environ.get("PATHEXT", "").split(os.pathsep)
+        # See if the given file matches any of the expected path extensions.
+        # This will allow us to short circuit when given "python.exe".
+        # If it does match, only test that one, otherwise we have to try
+        # others.
+        if any(cmd.lower().endswith(ext.lower()) for ext in pathext):
+            files = [cmd]
+        else:
+            files = [cmd + ext for ext in pathext]
+    else:
+        # On other platforms you don't have things like PATHEXT to tell you
+        # what file suffixes are executable, so just pass on cmd as-is.
+        files = [cmd]
+
+    seen = set()
+    for dir in path:
+        normdir = os.path.normcase(dir)
+        if not normdir in seen:
+            seen.add(normdir)
+            for thefile in files:
+                name = os.path.join(dir, thefile)
+                if _access_check(name, mode):
+                    return name
+    return None
+#end of "Which" function
+
+def gpg_path():
+    if sys.platform = "win32":
+        gpg = which('gpg2')
+    elif sys.platform = "darwin":
+        #joining all possible paths for location of mac binaries
+        BIN_PATHS = os.path.join(MAC_BIN_PATHS, LINUX_BIN_PATHS)
+        gpg = which('gpg2', path=BIN_PATHS)
+    elif sys.platform = "linux":
+        gpg = which('gpg2', path=LINUX_BIN_PATHS)
+    if gpg == "":
+        print("GPG2 could not be found! Do you have the correct permissions?")
+        sys.exit(1)
+    else:
+        return gpg
+
+def get_home():
+    if sys.platform == "win32":
+        return os.environ.get('APPDATA')
+    else:
+        return os.environ.get('HOME')
+
+def get_gpghome(usrhome):
+    if sys.platform == "win32":
+        return os.environ.get('GNUPGHOME', os.path.join(usrhome, 'gnupg'))
+    else:
+        return os.environ.get('GNUPGHOME', os.path.join(usrhome, '.gnupg'))
+
+def get_piushome(usrhome):
+    if sys.platform == "win32":
+        return os.path.join(usrhome, 'pius')
+    else:
+        #check to see if xdg home is a user env variable
+        #set it if it is to piushome
+        if os.environ.get('XDG_CONFIG_HOME') != "":
+            piushome = os.path.join(os.environ.get('XDG_CONFIG_HOME'), '.pius')
+        #long test to see if the path exists, is able to be accessed, and if it is a dir.
+        #then join the usrhome path with ,config/.pius and return it to the program.
+        elif os.path.exists(os.path.join(usrhome, ',config')) and os.access(os.path.join(usrhome,'.config'), mode) and os.path.isdir(os.path.join(usrhome,'.config')):
```

Comment:
This definitely isn't backwards compatible. Today we use `~/.pius/piusrc`, and if that's not there we fall back to `~/.pius`.

So in order of preference we probably want:

* $XDG_CONFIG_HOME/piusrc    # note that XDG_CONFIG_HOME defaults to `~/.config` so no need to add a `.config`
* $HOME/.pius/piusrc
* $HOME/.piusrc

Reply:
Ubuntu Server and newer versions of ubuntu doesn't use Xdesktop anymore for security reasons. They do have a .config file but no env variables for XDG_*

Reply:
So you're proposing the order be:
* $XDG_CONFIG_HOME/piusrc # note that XDG_CONFIG_HOME defaults to ~/.config so no need to add a .config
* $HOME/.config/piusrc
* $HOME/.pius/piusrc
* $HOME/.piusrc

? Hmmm. I don't love that. Just becuase they're moving to wayland doens't mean they shouldn't support common standards. How about we use their stuff if it's there **and** nothing else is there, so the order is:

* $XDG_CONFIG_HOME/piusrc # note that XDG_CONFIG_HOME defaults to ~/.config so no need to add a .config
* $HOME/.pius/piusrc
* $HOME/.piusrc
* $HOME/.config/piusrc

Reply:
Yeah that's failr, because it's just the default. IN that case:
 * $XDG_CONFIG_HOME/piusrc # note that XDG_CONFIG_HOME defaults to ~/.config so no need to add a .config
 * $HOME/.config/piusrc
 * $HOME/.pius/piusrc
* $HOME/.piusrc

Reply:
Then because of above I will have to change the checking order of things existing.